tools/schemacode/bidsschematools/render/text.py

```diff
@@ -370,37 +378,38 @@ def make_filename_template(
                 ]
 
             # Add extensions
-            extensions = [ext if ext != "*" else ".<extension>" for ext in group.extensions]
-            if len(extensions) >= n_dupes_to_combine:
-                # Combine exts when there are many, but keep JSON separate
-                if ".json" in extensions:
-                    extensions = [".<extension>", ".json"]
+            extension_groups = [
+                [".<extension>" if ext == "*" else ext]
+                if isinstance(ext, str)
+                else ([".<extension>"] if len(ext) >= n_dupes_to_combine else ext)
+                for ext in group.extensions
+            ]
+            for ext_group in sorted(extension_groups):
+                ext_headings = []
+                for extension in ext_group:
+                    # The glossary indexes by the extension identifier (niigz instead of .nii.gz),
+                    # but the rules reference the actual suffix string (.nii.gz instead of niigz),
+                    # so we need to look it up.
+                    key = ext_key_table.get(extension)
+                    if key:
+                        ext_headings.append(f"{key.lower()}-extensions")
+                    else:
+                        ext_headings.append("extension-common_principles")
+
+                extensions = utils.combine_extensions(
+                    ext_group,
+                    html_path=GLOSSARY_PATH + ".html",
+                    heading_lst=ext_headings,
+                    pdf_format=pdf_format,
+                )
+                if len(extensions) == 1:
+                    extension = extensions[0]
                 else:
-                    extensions = [".<extension>"]
-
-            ext_headings = []
-            for extension in extensions:
-                # The glossary indexes by the extension identifier (niigz instead of .nii.gz),
-                # but the rules reference the actual suffix string (.nii.gz instead of niigz),
-                # so we need to look it up.
-                key = ext_key_table.get(extension)
-                if key:
-                    ext_headings.append(f"{key.lower()}-extensions")
-                else:
-                    ext_headings.append("extension-common_principles")
-
-            extensions = utils.combine_extensions(
-                extensions,
-                html_path=GLOSSARY_PATH + ".html",
-                heading_lst=ext_headings,
-                pdf_format=pdf_format,
-            )
+                    extension = f"{lt}{'|'.join(extensions)}{gt}"
 
-            lines.extend(
-                f"\t\t\t{ent_string}_{suffix}{extension}"
-                for suffix in sorted(suffixes)
-                for extension in sorted(extensions)
-            )
+                lines.extend(
+                    f"\t\t\t{ent_string}_{suffix}{extension}" for suffix in sorted(suffixes)
+                )
```

Comment:
@Remi-Gau I found the large number of helper functions hard to reason about, so reverted large chunks and pushed this change. LMK what you think. Feel free to revert if you disagree.

I probably broke tests; I only tested by rendering.

Holiday today, so won't be on actively.

Reply:
haha

partly different coding style I think
I tend to prefer extracting a lot of small functions with their own tests, because it helps me better understand the whole by understanding the parts.
But I kinda agree that I did break things up here and added tests more as "scaffolding" while writing this: I guess that now that it runs and that our doc rendering acts as a system test.

Fine with me but let's check with @tsalo because he is one of us who has played a lot with the rendering code.

Reply:
Well-crafted small functions can be fine, but they can also take away flexibility as you now need to make your adjustments fit those functions, and if you need to modify one of those functions, it can affect something you're not trying to change.

For example, I started poking at this trying to see if we could turn `<.ext1|.ext2|.ext3>` into `.<ext1|ext2|ext3>` so long as all of the extensions started with a dot. The linking code means that `.ext1` is actually `<a href="...">.ext1</a>`, so that's no longer possible. That one's too deep to be worth digging out here. But when I rendered yours, I found that `.nii[.gz]` no longer looked the same, so that's when I reverted to see if I could make the grouping work with less refactoring.

Reply:
no worries
I trust you did for good reasons
as I said I mostly needed some training wheels: am fine with inlining those functions back in